### DISCUSSION TITLE
How to Properly Update Static Data in `CartesianChartModelProducer`?

### DISCUSSION THREAD

**Discussion by mak1nt0sh** [patrykandpatrick/vico#734] [Questions] [answered]
I'm using the Vico 2 to display a budget graph in a Cartesian chart, and the initial data loads and displays correctly. However, when new data is provided, the chart becomes empty. If I revert to the previous data, the chart displays correctly again. I suspect I'm not updating the chart data properly. The data needs to come platform independent, so I cannot use producer in `ViewModel`.

Here's my code:

```
@Composable
fun BudgetGraph(
    modifier: Modifier = Modifier,
    graphData: BudgetGraphData
) {
    val locale = LocalAppLocale.current

    val graphModeProducer = remember { CartesianChartModelProducer.build() }
    val xToDateMapKey = ExtraStore.Key<Map<Float, LocalDate>>()

    LaunchedEffect(graphData) {
        withContext(Dispatchers.Default) {
            val historyXEpochDays =
                graphData.historyGraphData.keys.associateBy { it.toEpochDay().toFloat() }

            val forecastXEpochDays =
                graphData.forecastGraphData.keys.associateBy { it.toEpochDay().toFloat() }

            graphModeProducer.tryRunTransaction {
                lineSeries {
                    series(x = historyXEpochDays.keys, y = graphData.historyGraphData.values)
                    series(x = forecastXEpochDays.keys, y = graphData.forecastGraphData.values)
                }
                updateExtras {
                    it[xToDateMapKey] = historyXEpochDays + forecastXEpochDays
                }
            }
        }
    }
}

data class BudgetGraphData(
    val limit: Float,
    val historyGraphData: Map<LocalDate, Float>,
    val forecastGraphData: Map<LocalDate, Float>
)
```

When new `graphData` is received, the chart does not display any data, but reverting to the previous `graphData` works fine.

What could be causing the chart to fail to update with new data? Are there specific steps or best practices I should follow to ensure the chart updates correctly with new data?

Thank you for your help!

**Comment by riley0521** [comment]
I am having the same problem right now, hope they can answer this 🙏

**Comment by patrickmichalik** [comment]
Hello, @mak1nt0sh! In this snippet, the only thing that sticks out to me as incorrect is that the `ExtraStore.Key` is recreated on every recomposition. `remember`, a top-level property, or a solution with a similar effect should be used. Please see [here](https://patrykandpatrick.com/vico/wiki/cartesian-charts/data#asynchrony-and-extras) for more. This mistake may lead to exceptions, so addressing it is important, but I wouldn’t expect it to cause the chart to go blank—the problem likely lies somewhere else. For good measure, though, I’d suggest starting by addressing the `ExtraStore.Key` issue and checking if that changes anything. If the chart continues to go blank, please share the following details:

1. a snippet showing how the `ExtraStore.Key` issue was resolved
2. the initial data and the data for which the chart goes blank
3. the return values of `CartesianChartModelProducer#tryRunTransaction` for both updates
4. a snippet showing how `CartesianChartHost` is used

For point 2, I’d suggest logging `graphData` from the `LaunchedEffect` lambda.

**Comment by mak1nt0sh** [comment]
Thanks for the suggestions, I figured it out that when `forecastGraphData` is empty the graph does not draw even if `historyGraphData` is present. I removed all `ExtraStore.Key` logic to simplify testing. If I wrap the series in `if(graphData.forecastGraphData.isNotEmpty())` everything works. 

Regarding the extras, I should do:

`val graphExtras = remember(graphData) { it[xToDateMapKey] = historyXEpochDays + forecastXEpochDays }` and simply use `updateExtras` with this? But the remember is redundant here because `LaunchedEffect` is called only when graphData changes. I'm sorry but the documentation you provided is a bit complex and hard to understand.

**Reply by Gowsky** [reply, marked as answer]
>when forecastGraphData is empty the graph does not draw even if historyGraphData is present

That's because empty series aren't allowed. The producer is throwing an exception, but due to the limitations of composable functions, the exception is swallowed. As a result you're seeing an empty composable function. We'll see if we can do something about the swallowed exception.

>Regarding the extras, I should do:

As @patrickmichalik mentioned it's about the key. You should either wrap it in `remember`, or move it outside of the composable. 
For example:
```kotlin
val xToDateMapKey = remember { ExtraStore.Key<Map<Float, LocalDate>>() }
```

As per the documentation:
>[ExtraStore.Key](https://patrykandpatrick.com/vico/api/2.0.0-alpha.20/core/com.patrykandpatrick.vico.core.common.data/-extra-store/-key/)s are compared by instance. Given the asynchronous context in which they’re used, it’s important not to recreate or swap them more often than appropriate. As a rule of thumb, all [ExtraStore.Key](https://patrykandpatrick.com/vico/api/2.0.0-alpha.20/core/com.patrykandpatrick.vico.core.common.data/-extra-store/-key/) instances used with a particular [CartesianChartModelProducer](https://patrykandpatrick.com/vico/api/2.0.0-alpha.20/core/com.patrykandpatrick.vico.core.cartesian.data/-cartesian-chart-model-producer/index.html) should be persisted at least as long as the [CartesianChartModelProducer](https://patrykandpatrick.com/vico/api/2.0.0-alpha.20/core/com.patrykandpatrick.vico.core.cartesian.data/-cartesian-chart-model-producer/index.html), and any given property’s [ExtraStore.Key](https://patrykandpatrick.com/vico/api/2.0.0-alpha.20/core/com.patrykandpatrick.vico.core.common.data/-extra-store/-key/) should remain unchanged over this period.

---
>I'm sorry but the documentation you provided is a bit complex and hard to understand.

Could you tell us what exactly is too hard to understand? Did you read the paragraph about the keys? Remembering instances of objects isn't specific to the library, but to Jetpack Compose. It's clearly mentioned that keys shouldn't be recreated.

**Reply by Gowsky** [reply]
Hi! [Vico 2.0.0-alpha.22](https://patrykandpatrick.com/vico/releases/2.0.0-alpha.22) fixes the exception swallowing.
